service/src/main/java/bio/terra/externalcreds/dataAccess/GoogleCloudStorageDAO.java

```diff
@@ -0,0 +1,122 @@
+
+@Repository
+@Slf4j
+public class GoogleCloudStorageDAO {
+
+  public static final int CONNECT_TIMEOUT_SECONDS = 20;
+  public static final int READ_TIMEOUT_SECONDS = 40;
+
+  private final Map<String, Storage> storageMap = new PassiveExpiringMap<>(15, TimeUnit.MINUTES);
+  /**
+   * It is important that the stream returned by this method be guaranteed to closed. Since this is
+   * a stream from IO, we need to make sure that the handle is closed. This can be done by wrapping
+   * the stream in a try-block once the stream is used for a terminal operation.
+   */
+  public Stream<Blob> streamBlobs(String googleProjectId, String bucketName) {
+    var storage = getStorage(googleProjectId);
+    final Storage.BlobListOption[] listOptions =
+        new Storage.BlobListOption[] {Storage.BlobListOption.currentDirectory()};
+    Iterable<Blob> blobs = storage.list(bucketName, listOptions).iterateAll();
+    return StreamSupport.stream(blobs.spliterator(), false);
+  }
+
+  public void writeEmptyBlob(Blob blob) {
+    try (var writer = blob.writer()) {
+      writer.write(ByteBuffer.wrap("".getBytes(StandardCharsets.UTF_8)));
+    } catch (IOException e) {
+      throw new ExternalCredsException(
+          String.format("Failed to write empty blob at %s", blob.getBlobId().toGsUtilUri()), e);
+    }
+  }
+
+  /**
+   * Write a {@link List<String>} to a GCS blob separated by newlines
+   *
+   * @param blobId blob id to write the lines to
+   * @param contentsToWrite contents to write to file
+   */
+  public void writeLinesToBlob(
+      String googleProjectId, BlobId blobId, List<String> contentsToWrite) {
+    var newLine = "\n".getBytes(StandardCharsets.UTF_8);
+    var storage = getStorage(googleProjectId);
+    Blob blob = storage.get(blobId);
+    if (blob == null) {
+      blob = storage.create(BlobInfo.newBuilder(blobId).build());
+    }
+    try (var writer = blob.writer()) {
+      contentsToWrite.forEach(
+          s -> {
+            try {
+              writer.write(ByteBuffer.wrap(s.getBytes(StandardCharsets.UTF_8)));
+              writer.write(ByteBuffer.wrap(newLine));
+            } catch (IOException e) {
+              throw new ExternalCredsException(
+                  String.format("Could not write to blob at %s. Line: %s", blobId.toGsUtilUri(), s),
+                  e);
+            }
+          });
+    } catch (IOException ex) {
+      throw new ExternalCredsException(
+          String.format("Could not write to blob at %s", blobId.toGsUtilUri()), ex);
+    }
+  }
+
+  public List<String> readLinesFromBlob(String googleProjectId, BlobId blobId) {
+    var storage = getStorage(googleProjectId);
+    return List.of(new String(storage.readAllBytes(blobId), StandardCharsets.UTF_8).split("\n"));
+  }
+
+  // Get a Storage with application-default credentials
+  @VisibleForTesting
+  public Storage getStorage(String googleProjectId) {
+    return storageMap.computeIfAbsent(
+        googleProjectId,
+        p -> {
```

Comment:
I am inclined to try to keep the logic inside a lambda as simple as possible.  How much is acceptable and how much is too much?  I'm not exactly sure, but this seems to fall on the "too much" side for me personally.  Multiple variable declarations, multiple steps, a try/catch, it's a lot to grok.  Not sure if this would be easier to test if it were broken or not, but I do think for maintainability, I prefer to break out logical chunks.  My take is that making something a method does not necessarily mean it needs to be reused, but that it should make the code generally more readable and maintainable as well.